rust/agama-software/src/model/registration.rs

```diff
@@ -18,47 +18,251 @@
+            self.base_product_specification(),
+            self.connect_params.clone(),
+        )?;
+        Ok(product)
+    }
+
+    fn base_product_specification(&self) -> suseconnect_agama::ProductSpecification {
+        Self::product_specification(&self.product, &self.version)
+    }
+
+    fn product_specification(id: &str, version: &str) -> suseconnect_agama::ProductSpecification {
+        suseconnect_agama::ProductSpecification {
+            identifier: id.to_string(),
+            arch: std::env::consts::ARCH.to_string(),
+            version: version.to_string(),
+        }
+    }
+
+    fn credentials_from_url(url: &str) -> Option<String> {
+        let url = Url::parse(url)
+            .inspect_err(|e| tracing::warn!("Could not parse the service URL: {e}"))
+            .ok()?;
+        url.query_pairs()
+            .find(|(k, _v)| k == "credentials")
+            .map(|(_k, v)| v.to_string())
+    }
 }
 
-#[derive(Clone, Serialize, Deserialize, utoipa::ToSchema)]
-pub struct RegistrationError {
-    /// ID of error. See dbus API for possible values
-    pub id: u32,
-    /// human readable error string intended to be displayed to user
-    pub message: String,
+/// A builder for a [Registration] object.
+///
+/// It is used to configure the build a registration object. It allows to configure
+/// the registration parameters like the product and version, the registration code,
+/// the e-mail, etc.
+/// [Registration] object.
+#[derive(Debug)]
+pub struct RegistrationBuilder {
+    root_dir: Utf8PathBuf,
+    product: String,
+    version: String,
+    code: Option<String>,
+    email: Option<String>,
+}
+
+impl RegistrationBuilder {
+    /// Creates a new builder.
+    ///
+    /// It receives the mandatory arguments for registering a system.
+    ///
+    /// * `root_dir`: root directory where libzypp configuration lives.
+    /// * `product`: product name (e.g., "SLES").
+    /// * `version`: product version (e.g., "16.1").
+    pub fn new(root_dir: Utf8PathBuf, product: &str, version: &str) -> Self {
+        RegistrationBuilder {
+            root_dir,
+            product: product.to_string(),
+            version: version.to_string(),
+            code: None,
+            email: None,
+        }
+    }
+
+    /// Sets the registration code to use.
+    ///
+    /// * `code`: registration code.
+    pub fn with_code(mut self, code: &str) -> Self {
+        self.code = Some(code.to_string());
+        self
+    }
+
+    /// Sets the e-mail associated to the registration.
+    pub fn with_email(mut self, email: &str) -> Self {
+        self.email = Some(email.to_string());
+        self
+    }
+
+    /// Registers the system and return a [Registration] object.
+    ///
+    /// It announces the system, gets the credentials and registers the base product.
+    ///
+    /// * `zypp`: zypp instance.
+    pub fn build(self, zypp: &zypp_agama::Zypp) -> RegistrationResult<Registration> {
```

Comment:
I am not sure if it is not confusing to call it Builder::build and in fact it do `registration`...so I kind of expect more method like `register` or `register_base_product`.